include/boost/redis/impl/sentinel_resolve_fsm.ipp

```diff
@@ -0,0 +1,184 @@
+
+      log_info(
+         st.logger,
+         "Trying to resolve the address of ",
+         st.cfg.sentinel.server_role == role::master ? "master" : "a replica of master",
+         " '",
+         st.cfg.sentinel.master_name,
+         "' using Sentinel");
+
+      // Try all Sentinels in order. Upon any errors, save the diagnostic and try with the next one.
+      // If none of them are available, print an error diagnostic and fail.
+      for (idx_ = 0u; idx_ < st.sentinels.size(); ++idx_) {
+         log_debug(st.logger, "Trying to contact Sentinel at ", st.sentinels[idx_]);
+
+         // Try to connect
+         BOOST_REDIS_YIELD(resume_point_, 1, st.sentinels[idx_])
+
+         // Check for cancellations
+         if (is_terminal_cancel(cancel_state)) {
+            log_debug(st.logger, "Sentinel resolve: cancelled (1)");
+            return system::error_code(asio::error::operation_aborted);
+         }
+
+         // Check for errors
+         if (ec) {
+            log_sentinel_error(st, idx_, "connection establishment error: ", ec);
+            continue;
+         }
+
+         // Execute the Sentinel request
+         log_debug(st.logger, "Executing Sentinel request at ", st.sentinels[idx_]);
+         st.sentinel_resp_nodes.clear();
+         BOOST_REDIS_YIELD(resume_point_, 2, sentinel_action::request())
+
+         // Check for cancellations
+         if (is_terminal_cancel(cancel_state)) {
+            log_debug(st.logger, "Sentinel resolve: cancelled (2)");
+            return system::error_code(asio::error::operation_aborted);
+         }
+
+         // Check for errors
+         if (ec) {
+            log_sentinel_error(st, idx_, "error while executing request: ", ec);
+            continue;
+         }
+
+         // Parse the response
+         sentinel_response resp;
+         ec = parse_sentinel_response(st.sentinel_resp_nodes, st.cfg.sentinel.server_role, resp);
+
+         if (ec) {
+            if (ec == error::resp3_simple_error || ec == error::resp3_blob_error) {
+               log_sentinel_error(st, idx_, "responded with an error: ", resp.diagnostic);
+            } else if (ec == error::resp3_null) {
+               log_sentinel_error(st, idx_, "doesn't know about the configured master");
+            } else {
+               log_sentinel_error(
+                  st,
+                  idx_,
+                  "error parsing response (maybe forgot to upgrade to RESP3?): ",
+                  ec);
+            }
+
+            continue;
+         }
+
+         // When asking for replicas, we might get no replicas
+         if (st.cfg.sentinel.server_role == role::replica && resp.replicas.empty()) {
+            log_sentinel_error(st, idx_, "the configured master has no replicas");
+            continue;
+         }
+
+         // Store the resulting address in a well-known place
+         if (st.cfg.sentinel.server_role == role::master) {
+            st.cfg.addr = resp.master_addr;
```

Comment:
IMO this assignment should be moved to the `run_op` right after `async_sentinel_resolve` resumes to make that function more readable. It would be more intuitive if `async_sentinel_resolve` complete with the value instead of changing the `connection_impl` state. Can we change the completion signature to `void(error_code, address)` to achieve that?

EDIT: We could also extend the completion signature with a third parameter which are the new sentinel addresses which is also use in run_op. This way the `async_sentinel_resolve` would not change any state in connection_impl.